modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/internal/servlet/taglib/BaseContainerTag.java

```diff
@@ -226,45 +262,56 @@ protected String processCssClasses(Set<String> cssClasses) {
 		return StringUtil.merge(cssClasses, StringPool.SPACE);
 	}
 
-	protected Map<String, Object> processData(Map<String, Object> data) {
-		data.put("cssClass", _cssClass);
-		data.put("id", _id);
-
-		return data;
-	}
-
 	protected int processEndTag() throws Exception {
 		JspWriter jspWriter = pageContext.getOut();
 
+		jspWriter.write("</");
+		jspWriter.write(_containerElement);
+		jspWriter.write(">");
+
 		String hydratedModuleName = getHydratedModuleName();
 
-		if (Validator.isNotNull(hydratedModuleName)) {
+		if (hydratedModuleName != null) {
 			NPMResolver npmResolver = NPMResolverProvider.getNPMResolver();
 
 			String moduleName = npmResolver.resolveModuleName(
 				hydratedModuleName);
 
+			String propsTransformer = null;
+
+			if (Validator.isNotNull(_propsTransformer)) {
+				propsTransformer = NPMResolvedPackageNameUtil.get(
+					pageContext.getServletContext()) + "/" + _propsTransformer;
+			}
+			else if (Validator.isNotNull(_defaultEventHandler)) {
+				propsTransformer = npmResolver.resolveModuleName(
+					"frontend-taglib-clay/DefaultEventHandlersPropsTransformer");
+			}
+
 			ComponentDescriptor componentDescriptor = new ComponentDescriptor(
-				moduleName, _id, new LinkedHashSet<>(), false);
+				moduleName, _id, new LinkedHashSet<>(), false,
+				propsTransformer);
 
 			ReactRenderer reactRenderer =
 				ReactRendererProvider.getReactRenderer();
 
 			reactRenderer.renderReact(
-				componentDescriptor, processData(new HashMap<>()), request,
+				componentDescriptor, prepareProps(new HashMap<>()), request,
 				jspWriter);
-		}
 
-		jspWriter.write("</");
-		jspWriter.write(_containerElement);
-		jspWriter.write(">");
+			jspWriter.write("</div>");
+		}
 
 		return EVAL_BODY_INCLUDE;
 	}
 
 	protected int processStartTag() throws Exception {
 		JspWriter jspWriter = pageContext.getOut();
 
+		if (getHydratedModuleName() != null) {
+			jspWriter.write("<div>");
```

Comment:
Still haven't implemented the `span` or `div` thing... wasn't sure about which API to use...
- `getHydratedContainerElement [div|span]`?
- `getHydratedContainerDisplay [inline|block]`?
- other?

Reply:
I lean towards `getHydratedContainerElement` as a low-level API that offers maximum control (for example, I might want to pick other elements for semantic purposes (eg. `<nav>`, `<section>`, `<article>` etc). Might even want to specify other attributes on the container (class etc).

Reply:
Ok, so I added it [here](https://github.com/liferay-frontend/liferay-portal/pull/2/commits/76942ed85d7cfd1e4dbd2e9f35fe87b7b98d91ed#diff-d9f615a14afa7fecd158f5d60c42b255R236) although I didn't expose this as an API yet... thinking about this now, we don't want the tag creator to control it but the tag user, so... I'll move this to a tag attribute rather than a tag method but avoid exposing it for now just to be sure this is the right approach. Hope that's okay.

Reply:
So, one should be able to do:

```html
<clay:dropdown-menu
    hydratedContainerElement="span"
    items="<%= items %>"
/>
```